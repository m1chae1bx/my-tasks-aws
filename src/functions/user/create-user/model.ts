import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface CreateUserRequest {
  body: {
    id: string;
    email: string;
    password: string;
    fullName: string;
    nickname: string;
  };
}

export const createUserRequestSchema: JSONSchemaType<CreateUserRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    body: {
      type: "object",
      properties: {
        id: {
          type: "string",
          maxLength: 100,
        },
        email: {
          type: "string",
        },
        password: {
          type: "string",
        },
        fullName: {
          type: "string",
        },
        nickname: {
          type: "string",
          maxLength: 100,
        },
      },
      required: ["id", "email", "password", "fullName", "nickname"],
    },
  },
  required: ["body"],
};

export const validateRequest = ajv.compile(createUserRequestSchema);
