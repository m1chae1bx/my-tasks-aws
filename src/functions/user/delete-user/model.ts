import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteUserRequest {
  pathParameters: {
    userId: string;
  };
  body: {
    password: string;
  };
}

export const deleteUserRequestSchema: JSONSchemaType<DeleteUserRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
          maxLength: 100,
        },
      },
      required: ["userId"],
    },
    body: {
      type: "object",
      properties: {
        password: {
          type: "string",
        },
      },
      required: ["password"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(deleteUserRequestSchema);
