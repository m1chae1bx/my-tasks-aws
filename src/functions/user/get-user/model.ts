import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface GetUserRequest {
  pathParameters: {
    userId: string;
  };
}

export const getUserRequestSchema: JSONSchemaType<GetUserRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
          maxLength: 100,
        },
      },
      required: ["userId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(getUserRequestSchema);
