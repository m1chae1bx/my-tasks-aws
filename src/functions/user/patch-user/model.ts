import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface PatchUserRequest {
  pathParameters: {
    userId: string;
  };
  body: {
    email?: string;
    fullName?: string;
    nickname?: string;
  };
}

export const patchUserRequestSchema: JSONSchemaType<PatchUserRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
          maxLength: 100,
        },
      },
      required: ["userId"],
    },
    body: {
      type: "object",
      properties: {
        email: {
          type: "string",
          nullable: true,
        },
        fullName: {
          type: "string",
          nullable: true,
        },
        nickname: {
          type: "string",
          maxLength: 100,
          nullable: true,
        },
      },
      additionalProperties: false,
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(patchUserRequestSchema);
