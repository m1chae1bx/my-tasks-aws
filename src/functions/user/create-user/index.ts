import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { User } from "/opt/nodejs/user.model";
import { genericErrorHandler, isAWSError } from "/opt/nodejs/util";

const emailUnavailableResponse = {
  message: "Email address is already registered",
  code: "emailUnavailable",
};

const usernameUnavailableResponse = {
  message: "Username is already taken",
  code: "usernameUnavailable",
};

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      body: event.body ? JSON.parse(event.body) : null,
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { id, email, password, fullName, nickname } = request.body;

    const userByEmail = await User.getByEmail(email);
    if (userByEmail) {
      return {
        statusCode: 409,
        body: JSON.stringify(emailUnavailableResponse),
      };
    } else {
      const userByUsername = await User.get(id);
      if (userByUsername) {
        return {
          // TODO create a Response object or use existing AWS if any?
          statusCode: 409,
          body: JSON.stringify(usernameUnavailableResponse),
        };
      } else {
        const newUser = new User(id, email, fullName, nickname);
        newUser.setPassword(password);
        await newUser.save();

        const response = {
          message: `User ${id} was created successfully`,
          id: id,
        };
        return {
          statusCode: 200,
          body: JSON.stringify(response),
        };
      }
    }
  } catch (error) {
    if (isAWSError(error) && error.code === "ConditionalCheckFailedException") {
      return {
        statusCode: 409,
        body: JSON.stringify(emailUnavailableResponse),
      };
    }

    return genericErrorHandler(
      error,
      `An error occurred while creating the user. Please try again later.`
    );
  }
};
