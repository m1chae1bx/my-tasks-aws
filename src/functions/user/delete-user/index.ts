import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { User } from "/opt/nodejs/user.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      body: event.body ? JSON.parse(event.body) : null,
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const userId = request.pathParameters.userId;
    const password = request.body.password;

    const user = await User.get(userId);
    if (!user) {
      return {
        statusCode: 404,
        body: JSON.stringify({
          message: "User not found",
          code: "notFound",
        }),
      };
    }
    if (!user.validatePassword(password)) {
      return {
        statusCode: 401,
        body: JSON.stringify({
          message: "Incorrect password",
          code: "incorrectPassword",
        }),
      };
    }

    await user.delete();

    return {
      statusCode: 200,
      body: JSON.stringify({
        message: `User ${userId} was deleted successfully`,
        code: "userDeleted",
      }),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      "An error occurred while deleting the user. Please try again later.",
      "deleteUserError"
    );
  }
};
