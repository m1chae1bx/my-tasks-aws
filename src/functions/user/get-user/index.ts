import { User } from "/opt/nodejs/user.model";
import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  const userId = event.pathParameters?.userId;
  if (!userId) {
    return {
      statusCode: 400,
      body: "Bad Request",
    };
  }

  try {
    const user = await User.get(userId);
    if (user) {
      delete user.hash;
      delete user.salt;
      return {
        statusCode: 200,
        body: JSON.stringify({ user }),
      };
    } else {
      return {
        statusCode: 404,
        body: JSON.stringify({ message: `User ${userId} was not found` }),
      };
    }
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while getting user ${userId}`
    );
  }
};
