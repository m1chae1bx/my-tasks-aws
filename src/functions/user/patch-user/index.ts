import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { User } from "/opt/nodejs/user.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  // TODO improve server side validation
  const userId = event.pathParameters?.userId;
  if (!userId) {
    return {
      statusCode: 400,
      body: "Bad Request",
    };
  }
  if (!event.body) {
    return {
      statusCode: 400,
      body: "Bad Request", // TODO: create a message property
    };
  }

  const body = JSON.parse(event.body);
  const { email, fullName, nickname } = body;
  const userPartial: Partial<User> = {
    id: userId,
    email,
    fullName,
    nickname,
  };

  try {
    await User.patch(userPartial);
    const response = {
      message: `User ${userId} was patched successfully`,
      id: userId,
    };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while patching user ${userId}`
    );
  }
};
