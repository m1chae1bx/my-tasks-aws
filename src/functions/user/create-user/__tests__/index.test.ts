import { APIGatewayProxyResult } from "aws-lambda";
import { handler } from "../index";
import { validateRequest } from "../model";
import {
  createUserEvent,
  testUser,
  validationErrors,
} from "../test-data/create-user-event";
import { UsernameUnavailableError } from "@libs/generic/errors";
import { User } from "@libs/user";
import { genericErrorHandler } from "@libs/generic/util";

jest.mock("../model", () => ({
  validateRequest: jest.fn(),
}));

jest.mock("@libs/user", () => ({
  ...jest.requireActual("@libs/user"),
  User: jest.fn(),
}));

jest.mock("@libs/generic/util", () => ({
  genericErrorHandler: jest.fn(),
}));

describe("create-user", () => {
  beforeAll(() => {
    User.getByEmail = jest.fn();
    User.get = jest.fn();
  });

  describe("happy path", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.getByEmail as jest.Mock).mockResolvedValueOnce(null);
      (User.get as jest.Mock).mockResolvedValueOnce(null);
      (User as unknown as jest.Mock).mockReturnValueOnce({
        setPassword: jest.fn(),
        save: jest.fn(),
      });
      result = await handler(createUserEvent);
    });

    it("should return a 200 response", () => {
      expect(result.statusCode).toBe(200);
    });

    it("should return the user id", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - invalid request", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = validationErrors;
      result = await handler(createUserEvent);
    });

    it("should return a 400 response", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return the validation errors", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - invalid request (no body)", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = validationErrors;
      result = await handler({ ...createUserEvent, body: null });
    });

    it("should return a 400 response", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return the validation errors", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - email unavailable", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.getByEmail as jest.Mock).mockResolvedValueOnce(testUser);

      result = await handler(createUserEvent);
    });

    it("should return a 409 response", () => {
      expect(result.statusCode).toBe(409);
    });

    it("should return an error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - username unavailable on get", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.getByEmail as jest.Mock).mockResolvedValueOnce(null);
      (User.get as jest.Mock).mockResolvedValueOnce(testUser);
      result = await handler(createUserEvent);
    });

    it("should return a 409 response", () => {
      expect(result.statusCode).toBe(409);
    });

    it("should return an error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - username unavailable error on catch", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.getByEmail as jest.Mock).mockResolvedValueOnce(null);
      (User.get as jest.Mock).mockResolvedValueOnce(null);
      (User as unknown as jest.Mock).mockReturnValueOnce({
        setPassword: jest.fn(),
        save: jest.fn().mockImplementationOnce(() => {
          throw new UsernameUnavailableError();
        }),
      });
      result = await handler(createUserEvent);
    });

    it("should return a 409 response", () => {
      expect(result.statusCode).toBe(409);
    });

    it("should return an error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - generic error", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.getByEmail as jest.Mock).mockResolvedValueOnce(null);
      (User.get as jest.Mock).mockResolvedValueOnce(null);
      (User as unknown as jest.Mock).mockReturnValueOnce({
        setPassword: jest.fn(),
        save: jest.fn().mockImplementationOnce(() => {
          throw new Error("generic error");
        }),
      });
      (genericErrorHandler as jest.Mock).mockClear();
      result = await handler(createUserEvent);
    });

    it("should pass error to genericErrorHandler", () => {
      expect((genericErrorHandler as jest.Mock).mock.calls).toMatchSnapshot();
    });
  });
});
