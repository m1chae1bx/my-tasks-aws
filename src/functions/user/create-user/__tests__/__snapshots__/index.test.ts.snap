// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-user happy path should return the user id 1`] = `"{\\"message\\":\\"User testUserId was created successfully\\",\\"id\\":\\"testUserId\\"}"`;

exports[`create-user sad path - email unavailable should return an error message 1`] = `"{\\"name\\":\\"EmailUnavailableError\\",\\"message\\":\\"Email address is already registered\\",\\"errorCode\\":\\"emailUnavailable\\"}"`;

exports[`create-user sad path - generic error should pass error to genericErrorHandler 1`] = `
Array [
  Array [
    [Error: generic error],
    "An error occurred while creating the user. Please try again later.",
  ],
]
`;

exports[`create-user sad path - invalid request should return the validation errors 1`] = `"{\\"message\\":\\"Bad request\\",\\"errors\\":[{\\"instancePath\\":\\"/body/password\\",\\"schemaPath\\":\\"#/properties/body/properties/password/minLength\\",\\"keyword\\":\\"minLength\\",\\"params\\":{\\"limit\\":8},\\"message\\":\\"must NOT have fewer than 8 characters\\"},{\\"instancePath\\":\\"/body/password\\",\\"schemaPath\\":\\"#/properties/body/properties/password/pattern\\",\\"keyword\\":\\"pattern\\",\\"params\\":{\\"pattern\\":\\"^(?=[^A-Za-z]*[A-Za-z])(?=[^0-9]*[0-9])(?=[^@#$%&!_:\\\\\\\\-.]*[@#$%&!_:\\\\\\\\-.]).*\\"},\\"message\\":\\"must match pattern \\\\\\"^(?=[^A-Za-z]*[A-Za-z])(?=[^0-9]*[0-9])(?=[^@#$%&!_:\\\\\\\\-.]*[@#$%&!_:\\\\\\\\-.]).*\\\\\\"\\"}]}"`;

exports[`create-user sad path - username unavailable error on catch should return an error message 1`] = `"{\\"name\\":\\"UsernameUnavailableError\\",\\"message\\":\\"Username is already taken\\",\\"errorCode\\":\\"usernameUnavailable\\"}"`;

exports[`create-user sad path - username unavailable on get should return an error message 1`] = `"{\\"name\\":\\"UsernameUnavailableError\\",\\"message\\":\\"Username is already taken\\",\\"errorCode\\":\\"usernameUnavailable\\"}"`;
